gum/backend-qnx/gumprocess-qnx.c

```diff
@@ -226,56 +317,69 @@ gum_process_enumerate_malloc_ranges (GumFoundMallocRangeFunc func,
                                      gpointer user_data)
 {
   /* Not implemented */
+  g_assert_not_reached ();
 }
 
 void
 gum_process_enumerate_modules (GumFoundModuleFunc func,
                                gpointer user_data)
 {
-  gchar * as_path;
   gint fd, res;
   gboolean carry_on = TRUE;
   procfs_mapinfo * mapinfos;
   gint num_mapinfos;
   procfs_debuginfo * debuginfo;
   gint i = 0;
+  struct _dl_phdr_internal ** handle;
+  struct _dl_phdr_internal * phdr;
 
-  as_path = g_strdup_printf ("/proc/%d/as", getpid ());
-  fd = open (as_path, O_RDONLY);
+  fd = open ("/proc/self/as", O_RDONLY);
   g_assert (fd != -1);
-  g_free (as_path);
 
   res = devctl (fd, DCMD_PROC_PAGEDATA, 0, 0, &num_mapinfos);
-  g_assert (res != 0);
+  g_assert (res == 0);
+
+  if (num_mapinfos == 0)
+    return;
 
   mapinfos = g_malloc (sizeof (procfs_mapinfo) * num_mapinfos);
-  debuginfo = g_malloc (sizeof (procfs_debug) + 0x100);
 
-  res = devctl (fd, DCMD_PROC_PAGEDATA, &mapinfos,
-      sizeof (procfs_mapinfo) * num_mapinfos, NULL);
-  g_assert (res != 0);
+  res = devctl (fd, DCMD_PROC_PAGEDATA, mapinfos,
+      sizeof (procfs_mapinfo) * num_mapinfos, &num_mapinfos);
+  g_assert (res == 0);
 
-  while (carry_on && i != num_mapinfos)
+  handle = dlopen(NULL, RTLD_NOW);
+
+  while (carry_on && (i != num_mapinfos))
   {
-    GumRangeDetails details;
+    GumModuleDetails details;
     GumMemoryRange range;
 
     range.base_address = mapinfos[i].vaddr;
     range.size = mapinfos[i].size;
 
     details.range = &range;
+    details.path = NULL;
 
-    debuginfo.vaddr = mapinfos[i].vaddr;
-    res = devctl (fd, DCMD_PROC_MAPDEBUG, debuginfo,
-        sizeof (procfs_debuginfo) + 0x100, NULL);
-    details.name = debuginfo.path;
+    phdr = *handle;
+    while (phdr && phdr->linkmap)
```

Comment:
Looks like a `for`-loop would make this read better? Also, please always be explicit and do `phdr != NULL` instead of `phdr` (except if `phdr` is a gboolean/bool).